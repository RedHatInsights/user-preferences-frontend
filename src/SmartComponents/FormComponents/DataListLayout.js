import React, { Fragment, Component } from 'react';
import {
    DataList,
    DataListItem,
    DataListItemRow,
    DataListItemCells,
    DataListCell
} from '@patternfly/react-core';
import PropTypes from 'prop-types';

class DataListRow extends Component {
    state = {
        hasError: false
    }

    componentDidCatch(_error, info) {
        this.setState({ hasError: info });
    }

    render() {
        const { fieldsKey, label, fields, formOptions } = this.props;
        const { hasError } = this.state;
        return <DataListItemRow>
            <DataListItemCells dataListCells={ [
                <DataListCell isFilled={ false } className="pref-c-title pref-u-bold" key="title">
                    { fieldsKey === 0 ? label : '' }
                </DataListCell>,
                <DataListCell isFilled key={ `${fieldsKey}-content` }>
                    { hasError ?
                        'Error while mapping fields, please check your schema if it has valid component types' :
                        formOptions.renderForm(fields, formOptions)
                    }
                </DataListCell>
            ] } />
        </DataListItemRow>;
    }
}

DataListRow.propTypes = {
    fieldsKey: PropTypes.string,
    fields: PropTypes.array,
    FieldProvider: PropTypes.any,
    formOptions: PropTypes.any,
    label: PropTypes.node,
    name: PropTypes.string,
    validate: PropTypes.any,
    FormSpyProvider: PropTypes.any
};

// eslint-disable-next-line no-unused-vars
const DataListLayout = ({ sections, label, name, FormSpyProvider, FieldProvider, formOptions, validate, ...props }) => (
    <DataList aria-label={ label || name } { ...props }>
        {sections.map(({ label, fields }, key) => {
            const actualFields = Array.isArray(fields) ? fields : [ fields ];
            return (
                actualFields.length === 0 ? <Fragment /> : <DataListItem name={ name } key={ key } aria-labelledby="simple-item1">
                    {actualFields.map(({ fields: fieldsToRender }, fieldsKey) => {
                        return (
                            <DataListRow key={ fieldsKey }
                                fieldsKey={ fieldsKey }
                                label={ label }
                                fields={ fieldsToRender }
                                formOptions={ formOptions }
                            />
                        );
                    })}
                </DataListItem>
            );
        })}
    </DataList>
);

DataListLayout.propTypes = {
    sections: PropTypes.array,
    FieldProvider: PropTypes.any,
    formOptions: PropTypes.any,
    label: PropTypes.node,
    name: PropTypes.string,
    validate: PropTypes.any,
    FormSpyProvider: PropTypes.any
};

export default DataListLayout;
