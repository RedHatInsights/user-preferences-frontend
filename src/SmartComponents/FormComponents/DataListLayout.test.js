import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import DataListLayout from './DataListLayout';

describe('DataListLayout checkbox tests', () => {

    it('should render correctly', () => {
        const wrapper = mount(
            <DataListLayout
                label="test label"
                sections={ [{ label: 'test', fields: [{ fields: []}]}] }
                formOptions={ {
                    renderForm: () => 'test'
                } }
            />
        );
        expect(toJson(wrapper.find('DataListLayout'))).toMatchSnapshot();
    });
});
