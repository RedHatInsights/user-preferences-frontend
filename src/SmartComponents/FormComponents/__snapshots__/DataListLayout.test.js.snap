// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataListLayout checkbox tests should render correctly 1`] = `
<DataListLayout
  __versions={
    Object {
      "final-form": "4.19.1",
      "react-final-form": "6.4.0",
    }
  }
  clearedValue={true}
  dirty={false}
  dirtyFields={Object {}}
  dirtyFieldsSinceLastSubmit={Object {}}
  dirtySinceLastSubmit={false}
  errors={Object {}}
  form={
    Object {
      "batch": [Function],
      "blur": [Function],
      "change": [Function],
      "destroyOnUnregister": false,
      "focus": [Function],
      "getFieldState": [Function],
      "getRegisteredFields": [Function],
      "getState": [Function],
      "initialize": [Function],
      "isValidationPaused": [Function],
      "mutators": Object {},
      "pauseValidation": [Function],
      "registerField": [Function],
      "reset": [Function],
      "resetFieldState": [Function],
      "resumeValidation": [Function],
      "setConfig": [Function],
      "submit": [Function],
      "subscribe": [Function],
    }
  }
  formOptions={
    Object {
      "renderForm": [Function],
    }
  }
  handleSubmit={[Function]}
  hasSubmitErrors={false}
  hasValidationErrors={false}
  invalid={false}
  label="test label"
  modified={
    Object {
      "undefined": false,
    }
  }
  pristine={true}
  sections={
    Array [
      Object {
        "fields": Array [
          Object {
            "fields": Array [],
          },
        ],
        "label": "test",
      },
    ]
  }
  submitFailed={false}
  submitSucceeded={false}
  submitting={false}
  touched={
    Object {
      "undefined": false,
    }
  }
  valid={true}
  validating={false}
  values={Object {}}
  visited={
    Object {
      "undefined": false,
    }
  }
>
  <Component
    __versions={
      Object {
        "final-form": "4.19.1",
        "react-final-form": "6.4.0",
      }
    }
    aria-label="test label"
    clearedValue={true}
    dirty={false}
    dirtyFields={Object {}}
    dirtyFieldsSinceLastSubmit={Object {}}
    dirtySinceLastSubmit={false}
    errors={Object {}}
    form={
      Object {
        "batch": [Function],
        "blur": [Function],
        "change": [Function],
        "destroyOnUnregister": false,
        "focus": [Function],
        "getFieldState": [Function],
        "getRegisteredFields": [Function],
        "getState": [Function],
        "initialize": [Function],
        "isValidationPaused": [Function],
        "mutators": Object {},
        "pauseValidation": [Function],
        "registerField": [Function],
        "reset": [Function],
        "resetFieldState": [Function],
        "resumeValidation": [Function],
        "setConfig": [Function],
        "submit": [Function],
        "subscribe": [Function],
      }
    }
    formOptions={
      Object {
        "renderForm": [Function],
      }
    }
    handleSubmit={[Function]}
    hasSubmitErrors={false}
    hasValidationErrors={false}
    invalid={false}
    label="test label"
    modified={
      Object {
        "undefined": false,
      }
    }
    pristine={true}
    sections={
      Array [
        Object {
          "fields": Array [
            Object {
              "fields": Array [],
            },
          ],
          "label": "test",
        },
      ]
    }
    submitFailed={false}
    submitSucceeded={false}
    submitting={false}
    touched={
      Object {
        "undefined": false,
      }
    }
    valid={true}
    validating={false}
    values={Object {}}
    visited={
      Object {
        "undefined": false,
      }
    }
  >
    <ul
      __versions={
        Object {
          "final-form": "4.19.1",
          "react-final-form": "6.4.0",
        }
      }
      aria-label="test label"
      className="pf-c-data-list"
      clearedValue={true}
      dirty={false}
      dirtyFields={Object {}}
      dirtyFieldsSinceLastSubmit={Object {}}
      dirtySinceLastSubmit={false}
      errors={Object {}}
      form={
        Object {
          "batch": [Function],
          "blur": [Function],
          "change": [Function],
          "destroyOnUnregister": false,
          "focus": [Function],
          "getFieldState": [Function],
          "getRegisteredFields": [Function],
          "getState": [Function],
          "initialize": [Function],
          "isValidationPaused": [Function],
          "mutators": Object {},
          "pauseValidation": [Function],
          "registerField": [Function],
          "reset": [Function],
          "resetFieldState": [Function],
          "resumeValidation": [Function],
          "setConfig": [Function],
          "submit": [Function],
          "subscribe": [Function],
        }
      }
      formOptions={
        Object {
          "renderForm": [Function],
        }
      }
      handleSubmit={[Function]}
      hasSubmitErrors={false}
      hasValidationErrors={false}
      invalid={false}
      label="test label"
      modified={
        Object {
          "undefined": false,
        }
      }
      pristine={true}
      sections={
        Array [
          Object {
            "fields": Array [
              Object {
                "fields": Array [],
              },
            ],
            "label": "test",
          },
        ]
      }
      submitFailed={false}
      submitSucceeded={false}
      submitting={false}
      touched={
        Object {
          "undefined": false,
        }
      }
      valid={true}
      validating={false}
      values={Object {}}
      visited={
        Object {
          "undefined": false,
        }
      }
    >
      <Component
        aria-labelledby="simple-item1"
        key="0"
      >
        <li
          aria-labelledby="simple-item1"
          className="pf-c-data-list__item"
          id=""
        >
          <DataListRow
            fields={Array []}
            fieldsKey={0}
            formOptions={
              Object {
                "renderForm": [Function],
              }
            }
            key=".$0-0"
            label="test"
            rowid="simple-item1"
          >
            <Component>
              <div
                className="pf-c-data-list__item-row"
              >
                <Component
                  dataListCells={
                    Array [
                      <Unknown
                        className="pref-c-title pref-u-bold"
                        isFilled={false}
                      >
                        test
                      </Unknown>,
                      <Unknown
                        isFilled={true}
                      />,
                    ]
                  }
                  key=".0"
                  rowid=""
                >
                  <div
                    className="pf-c-data-list__item-content"
                  >
                    <Component
                      className="pref-c-title pref-u-bold"
                      isFilled={false}
                      key="0-title"
                    >
                      <div
                        className="pf-c-data-list__cell pf-m-no-fill pref-c-title pref-u-bold"
                      >
                        test
                      </div>
                    </Component>
                    <Component
                      isFilled={true}
                      key="0-content"
                    >
                      <div
                        className="pf-c-data-list__cell"
                      />
                    </Component>
                  </div>
                </Component>
              </div>
            </Component>
          </DataListRow>
        </li>
      </Component>
    </ul>
  </Component>
</DataListLayout>
`;
